supervisor/host/logs.py

```diff
@@ -0,0 +1,138 @@
+from aiohttp.hdrs import ACCEPT, RANGE
+
+from ..coresys import CoreSys, CoreSysAttributes
+from ..exceptions import HostLogError, HostNotSupportedError
+from .const import PARAM_BOOT_ID, PARAM_SYSLOG_IDENTIFIER, LogFormat
+
+_LOGGER: logging.Logger = logging.getLogger(__name__)
+
+SYSTEMD_JOURNAL_GATEWAYD_SOCKET: Path = Path("/run/systemd-journal-gatewayd.sock")
+
+# From systemd catalog for message IDs (`journalctl --dump-catalog``)
+# -- b07a249cd024414a82dd00cd181378ff
+# Subject: System start-up is now complete
+# Defined-By: systemd
+BOOT_IDS_QUERY = {"MESSAGE_ID": "b07a249cd024414a82dd00cd181378ff"}
+
+
+class LogsControl(CoreSysAttributes):
+    """Handle systemd-journal logs."""
+
+    def __init__(self, coresys: CoreSys):
+        """Initialize host power handling."""
+        self.coresys: CoreSys = coresys
+        self._profiles: set[str] = set()
+        self._boot_ids: list[str] = []
+        self._identifiers: list[str] = []
+
+    @property
+    def available(self) -> bool:
+        """Return True if Unix socket to systemd-journal-gatwayd is available."""
+        return SYSTEMD_JOURNAL_GATEWAYD_SOCKET.is_socket()
+
+    @property
+    def boot_ids(self) -> list[str]:
+        """Get boot IDs from oldest to newest."""
+        return self._boot_ids
+
+    @property
+    def identifiers(self) -> list[str]:
+        """Get syslog identifiers."""
+        return self._identifiers
+
+    def get_boot_id(self, offset: int = 0):
+        """
+        Get ID of a boot by offset.
+
+        Current boot is offset = 0, negative numbers go that many in the past.
+        Positive numbers count up from the oldest boot.
+        """
+        if not self.boot_ids:
+            raise HostLogError(
+                "Unable to obtain boot IDs from host, check logs for errors"
+            )
+
+        offset -= 1
+        if offset >= len(self.boot_ids) or abs(offset) > len(self.boot_ids):
+            raise ValueError(f"Logs only contain {len(self.boot_ids)} boots")
+
+        return self.boot_ids[offset]
+
+    async def update(self) -> None:
+        """Cache boot and identifier information."""
+        try:
+            async with self.journald_logs(
+                path=f"/fields/{PARAM_SYSLOG_IDENTIFIER}",
+                timeout=ClientTimeout(total=20),
+            ) as resp:
+                self._identifiers = (await resp.text()).split("\n")
+        except (ClientError, TimeoutError) as err:
+            raise HostLogError(
+                "Could not get a list of syslog identifiers from systemd-journal-gatewayd",
+                _LOGGER.error,
+            ) from err
+
+        if self.boot_ids:
+            # Doesn't change without a reboot, no reason to query again once cached
+            return
+
+        try:
+            async with self.journald_logs(
+                params=BOOT_IDS_QUERY,
+                accept=LogFormat.JSON,
+                timeout=ClientTimeout(total=20),
+            ) as resp:
```

Comment:
I am still slightly worried what this might do in practice. From tests on Raspberry Pi 3/4 it takes up to 2s to complete. It is *much* faster than `journalctl --list-boots`, so that is nice. Let's see how this is going.

Reply:
That said, I do prefer this solution over keeping track ourselfs... So definitely a :+1: from my side :smile:

Reply:
Switched over to loading boot IDs and identifiers only on demand. So identifiers will be fetching on each call to `/host/logs/identifiers` and not otherwise. Caching was removed here, user will get the current list of identifiers each time (figured repeated calls in a short interval probably wasn't common enough to warrant caching).

Boot IDs will be fetched on first call to `/host/logs/boots` or `/host/logs/boots/{boot_id}` where `boot_id` is an offset number rather then an actual boot identifier. This will be cached on first use and then the query won't be run again until next restart of supervisor.

Hopefully that will be good. Then user will only feel the perf hit here if they actually need this feature.